bundles/remote_services/topology_manager/src/topology_manager.c

```diff
@@ -380,61 +349,64 @@ celix_status_t topologyManager_exportScopeChanged(void *handle, char *filterStr)
 	}
 
 	// add already exported services to new rsa
-	if (celixThreadMutex_lock(&manager->exportedServicesLock) == CELIX_SUCCESS) {
-		hash_map_iterator_pt exportedServicesIterator = hashMapIterator_create(manager->exportedServices);
-		int size = hashMap_size(manager->exportedServices);
-		service_reference_pt *srvRefs = (service_reference_pt *) calloc(size, sizeof(service_reference_pt));
-		char **srvIds = (char **) calloc(size, sizeof(char*));
-		int nrFound = 0;
-
-		found = false;
-
-		while (hashMapIterator_hasNext(exportedServicesIterator)) {
-			hash_map_entry_pt entry = hashMapIterator_nextEntry(exportedServicesIterator);
-			service_reference_pt reference = hashMapEntry_getKey(entry);
-			reg = NULL;
-			serviceReference_getServiceRegistration(reference, &reg);
-			if (reg != NULL) {
-				props = NULL;
-				serviceRegistration_getProperties(reg, &props);
-				status = filter_match(filter, props, &found);
-				if (found) {
-					srvRefs[nrFound] = reference;
-					serviceReference_getProperty(reference, OSGI_FRAMEWORK_SERVICE_ID, &serviceId);
-					srvIds[nrFound++] = (char*)serviceId;
-				}
+	celixThreadMutex_lock(&manager->lock);
+
+	hash_map_iterator_pt exportedServicesIterator = hashMapIterator_create(manager->exportedServices);
+	int size = hashMap_size(manager->exportedServices);
+	service_reference_pt *srvRefs = (service_reference_pt *) calloc(size, sizeof(service_reference_pt));
+	char **srvIds = (char **) calloc(size, sizeof(char*));
+	int nrFound = 0;
+
+	found = false;
+
+	while (hashMapIterator_hasNext(exportedServicesIterator)) {
+		hash_map_entry_pt entry = hashMapIterator_nextEntry(exportedServicesIterator);
+		service_reference_pt reference = hashMapEntry_getKey(entry);
+		reg = NULL;
+		serviceReference_getServiceRegistration(reference, &reg);
+		if (reg != NULL) {
+			props = NULL;
+			serviceRegistration_getProperties(reg, &props);
+			status = filter_match(filter, props, &found);
+			if (found) {
+				srvRefs[nrFound] = reference;
+				serviceReference_getProperty(reference, OSGI_FRAMEWORK_SERVICE_ID, &serviceId);
+				srvIds[nrFound++] = (char*)serviceId;
 			}
 		}
+	}
 
-		hashMapIterator_destroy(exportedServicesIterator);
-		celixThreadMutex_unlock(&manager->exportedServicesLock);
+	hashMapIterator_destroy(exportedServicesIterator);
 
-		if (nrFound > 0) {
-			for (int i = 0; i < nrFound; i++) {
-				// Question: can srvRefs become invalid meanwhile??
-				const char* export = NULL;
-				serviceReference_getProperty(srvRefs[i], (char *) OSGI_RSA_SERVICE_EXPORTED_INTERFACES, &export);
+	if (nrFound > 0) {
+		for (int i = 0; i < nrFound; i++) {
+			// Question: can srvRefs become invalid meanwhile??
```

Comment:
Well, to be honest I am not sure. Service references are not protected and that is why there are not directly used anymore in the later Celix bundles / Celix api (celix_ prefixed api). That being, said the currently Celix also updates all service register/unregister events through the Celix event thread and I think this is called in the Celix event thread, which means the service reference cannot become invalid. 

Ideally (but I think out of scope for this pull request) the services (endpoint listeners, RSAs, services marked for export, etc) are tracked using the celix_bundleContext_trackServices calls instead of the current deprecated usage of directly working on a service tracker with service references.

Reply:
If we extend managed tracker (created by `celix_bundleContext_trackServices`) API to support CELIX_SERVICE_USE_DIRECT, as already discussed here: https://github.com/apache/celix/pull/399#discussion_r816607673, then a lot of external synchronizations, such as  `discovery->discoveredServicesMutex`, can be eliminated.